module/zfs/zfs_vfsops.c

```diff
@@ -2197,6 +2205,12 @@ zfs_resume_fs(zfsvfs_t *zfsvfs, dsl_dataset_t *ds)
 	mutex_enter(&zfsvfs->z_znodes_lock);
 	for (zp = list_head(&zfsvfs->z_all_znodes); zp;
 	    zp = list_next(&zfsvfs->z_all_znodes, zp)) {
+		/* see comment in zfs_suspend_fs() */
+		if (zp->z_suspended) {
+			iput(ZTOI(zp));
+			zp->z_suspended = B_FALSE;
+		}
```

Comment:
Calling `iput()` here directly isn't safe since the inode can be freed by `iput()->iput_final()` when the last reference is dropped.  As part of that destroy/free `->zfs_inode_destroy()` is called which will take the `z_znodes_lock` lock resulting in a deadlock.  You're going to need to call `zfs_iput_async()` after the `zfs_rezget()` instead.  This ensures the `iput()` is handled asynchronously by a taskq thread when it's the last reference.